lib/IRGen/GenStruct.cpp

```diff
@@ -337,6 +339,43 @@ namespace {
                                    const ClangFieldInfo &field) const {
       llvm_unreachable("non-fixed field in Clang type?");
     }
+
+    void initializeWithCopy(IRGenFunction &IGF, Address destAddr,
+                            Address srcAddr, SILType T, bool isOutlined) const {
+      const clang::CXXConstructorDecl *copyConstructor;
+      for (auto member : T.getStructOrBoundGenericStruct()->getMembers()) {
+        if (auto ctor =
+                dyn_cast<clang::CXXConstructorDecl>(member->getClangDecl())) {
+          if (ctor->isCopyConstructor()) {
+            copyConstructor = ctor;
+            break;
+          }
+        }
+      }
+      // TODO: here we should look for the copy asignment operator, otherwise
+      // we'll emit a memcpy when we should call a "real" copy constructor.
```

Comment:
Yeah, it's not super clear. Essentially I'm saying that sometimes clang defines a copy assignment operator and _not_ a copy constructor. We should be able to look for both and pick whichever is defined. 

I was able to fix this, though, by copying the sema code from #30630. (In the future, not this patch, we should also be able to look for copy assignment operators, though).

Reply:
Additionally, I agree with the second bullet.

Reply:
> In the future, not this patch, we should also be able to look for copy assignment operators, though

In which situations would we want to do this?

Unless the copy constructor is deleted, we can always get Sema to define the copy constructor if the user didn't provide a definition. Using that will always be more efficient, and arguably more correct, than using the combination of default constructor and assignment operator. Besides, if the class doesn't define a default constructor, we won't even be able to use the latter option.

Reply:
Fair enough. I think you're right. My thinking was that we might want to use the copy assignment operator when the user has defined that and not defined a copy constructor (in which case, I think it would be implicitly deleted).

Reply:
Yes, I believe that's true -- essentially, it guards against violations of the rule of five. In other words, we're dealing with a questionably defined class, and so it seems safer simply not to import it rather than trying (and possibly failing) to do the right thing.